### DISCUSSION TITLE
Could r25b support neon intrinsics like vcge_f16?

### DISCUSSION THREAD

**Discussion by xxxxxxLD** [android/ndk#1776] [Q&A]
I want to use r25b ndk to build neon intrinsics code , and some errors occur, it seems like 'vcge_f16' is an undeclared identifier，and in Android.mk I opened the '-mfloat-abi=softfp' ,What is the best approach for building the neon instrinsics?

**Comment by pirama-arumuga-nainar** [comment]
This particular intrinsic is only available with the fp16fml AArch64 extension (see below on how to pass this feature flag to Clang).  The caveat is that the fp16 extension is not available in all SOCs and you'd need runtime checks (e.g. ifuncs) to fallback to non-intrinsic code.

```
$ cat test.c
#include <arm_neon.h>

uint16x4_t foo(float16x4_t a, float16x4_t b) {
  return vcge_f16(a, b);
}
$ NDK/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android24-clang -c test.c -march=armv8+fp16
// passes
$ NDK/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android24-clang -c test.c
test.c:4:10: warning: implicit declaration of function 'vcge_f16' is invalid in C99 [-Wimplicit-function-declaration]
  return vcge_f16(a, b);
         ^
test.c:4:10: error: returning 'int' from a function with incompatible result type 'uint16x4_t' (vector of 4 'uint16_t' values)
  return vcge_f16(a, b);
         ^~~~~~~~~~~~~~
1 warning and 1 error generated.
```

**Reply by xxxxxxLD** [reply]
Thanks, and what about 'vcmla_rot270_f32' neon intrinsics? How to pass the complex arithmentic feature flag to Clang?

**Comment by pirama-arumuga-nainar** [comment]
The flag for this intrinsic would be `-march=armv8.3a`.

In general, these intrinsics are only declared if the corresponding feature/arch is enabled for the selected target.  Here's my approach/steps to find which flags enable an intrinsic:
- Look for the intrinsic in $NDK/toolchains/llvm/prebuilt/linux-x86_64/lib64/clang/14.0.6/include/arm_neon.h or other files in that directory.
- Find the `#if` guard that controls the declaration.  For `vcmla_rot270_f32`, that is `__ARM_FEATURE_COMPLEX`.
- Look at clang's target-features [test](https://github.com/llvm/llvm-project/blob/main/clang/test/Preprocessor/aarch64-target-features.c) to find which options define this guard.  `__ARM_FEATURE_COMPLEX` is defined for `armv8.3-a` and newer.

This is quite obtuse way of finding this info though - I wish hardware vendors (or the compiler) make it easier to discover.

**Reply by xxxxxxLD** [reply]
Thanks a lot.It's quite inconvenient to add each corresponding flags for intrinsics.

**Reply by enh-google** [reply]
> Thanks a lot.It's quite inconvenient to add each corresponding flags for intrinsics.

yes, but that's working as intended --- the whole point here is that you _probably don't have these features available_. you're going to need to write a bunch of extra code to check at runtime whether you're on a new enough cpu, and only load this code if so (and load code built with standard flags otherwise).

(if the reason you know you're safe using these features is because you're not actually writing an app, and you're working on the OS itself, "you're doing it wrong" :-) --- you need to update your device's BoardConfig.mk to state what hardware you're actually shipping, and then you'll get this kind of thing for free.)
